### ISSUE TITLE
Load custom DSDL

### ISSUE THREAD

**Issue by antoinealb** [UAVCAN/gui_tool#11]
Hi,
I am trying to debug code that uses custom messages on UAVCAN, and I would like a way to load custom DSDL types in the GUI. Currently I did an ugly patch to do it (see below) but I would be interested in implementing a more complete feature if needed by others.

```diff
commit 2bc955ece0d4c699ac75755e6381cd7c2e1c1f09
Author: Antoine Albertelli <antoine@antoinealb.net>
Date:   Wed Mar 29 15:59:26 2017 +0200

    Ugly hack to load custom datatypes

diff --git a/uavcan_gui_tool/widgets/bus_monitor/window.py b/uavcan_gui_tool/widgets/bus_monitor/window.py
index 8b755ae..a5803d0 100644
--- a/uavcan_gui_tool/widgets/bus_monitor/window.py
+++ b/uavcan_gui_tool/widgets/bus_monitor/window.py
@@ -11,6 +11,7 @@ import time
 import os
 from functools import partial
 import uavcan
+uavcan.load_dsdl('.')
 from uavcan.driver import CANFrame
 from PyQt5.QtWidgets import QMainWindow, QHeaderView, QLabel, QSplitter, QSizePolicy, QWidget, QHBoxLayout, \
     QPlainTextEdit, QDialog, QVBoxLayout, QMenu, QAction
```

**Comment by pavel-kirienko** [comment]
Thanks Antoine, it is needed by others. Please do implement that. ;)

There have been attempts to implement this feature in the past, but IIRC they ended up being unsuccessful.

How are you planning to implement that exactly? Are you perhaps going to add an environment variable specifying the look up paths, or is it going to be a GUI element where the user can specify the look up path manually? I think both approaches make sense.

**Comment by antoinealb** [comment]
I am not sure of the best way to implement it. I am not very familiar with (Py)Qt, so I think an environment variable would probably be the easiest for me, but not the most user friendly.

Also do you think having it in the bus monitor is enough or is there any other place where it could be useful ?

**Comment by pavel-kirienko** [comment]
> environment variable would probably be the easiest for me, but not the most user friendly.

Fine. Do you think it makes sense to implement that on the level of PyUAVCAN? We have a line here where the default DSDL definitions are loaded: https://github.com/UAVCAN/pyuavcan/blob/master/uavcan/__init__.py#L183. We could extend that code so that it queries the environment variable and loads DSDL from the locations specified in the variable.

> Also do you think having it in the bus monitor is enough or is there any other place where it could be useful ?

Having it only in the bus monitor is definitely not enough. There are other parts of the application, such as the plotter, the subscriber tool, and the console, where the user may need access to the custom definitions. Implementing the support for environment variable on the level of PyUAVCAN will take care of that.

**Comment by pavel-kirienko** [comment]
I have just discovered an [undocumented feature in PyUAVCAN](https://github.com/UAVCAN/pyuavcan/blob/master/uavcan/__init__.py#L123-L126). Have a look:

```python
paths = [os.path.join(dsdl_path, "uavcan")] + paths
custom_path = os.path.join(os.path.expanduser("~"), "uavcan_vendor_specific_types")
if os.path.isdir(custom_path):
    paths += [f for f in [os.path.join(custom_path, f) for f in os.listdir(custom_path)]
              if os.path.isdir(f)]
```

Missing documentation entries have been added to the [GUI Tool](http://uavcan.org/GUI_Tool/User_guide/#using-vendor-specific-dsdl-definitions) and [PyUAVCAN](http://uavcan.org/Implementations/Pyuavcan/Tutorials/2._Basic_usage/#using-vendor-specific-dsdl-definitions) docs.

That said, we still very much want the capability to point the location where to load the custom DSDL from via GUI.

**Comment by antoinealb** [comment]
Interesting. It should be good enough for my application, thanks!

**Comment by pavel-kirienko** [comment]
Solved in #27
